### ISSUE TITLE
Kafka portals broker mapping

### ISSUE THREAD

**Issue by hairyhum** [build-trust/ockam#4011]
Kafka binary protocol instructs clients to connect to different brokers (kafka servers) to address different topic partitions. In order to discover which broker is used for which partition, the metadata request is used: https://kafka.apache.org/24/protocol.html#The_Messages_Metadata

This protocol response assigns a `node_id` to each broker and uses it as a reference in the partitions information.

Since application client cannot connect directly to kafka brokers, kafka portal should replace them in the metadata response with local addresses (with different ports) and save the mapping between local ports (which application client is going to connect to) and remote brokers.
Then it needs to add information about which remote broker it's addressing in every request using Wrapped kafka request/response format described in https://github.com/build-trust/ockam/issues/3963

Metadata request:
![Kafka end-to-end - Broker mapping metadata](https://user-images.githubusercontent.com/784180/210109850-8cbe2bf8-b714-4286-a2d8-38e589c452ab.jpg)

After metadata request:
![Kafka end-to-end - Broker mapping request](https://user-images.githubusercontent.com/784180/210109866-7594aace-79e6-4143-a24c-596b8a56b7f9.jpg)

**Comment by mszpakowski** [comment]
Do we really need that? I tested it and it seems to have been reconciled under the hood.

```elixir
> :brod.get_metadata([{"<hostname>.us-east-1.aws.confluent.cloud", 9092}], ["test"], config)                                                                             

{:ok,
 %{
   brokers: [
     %{
       host: "b0-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 0,
       port: 9092,
       rack: "use1-az1"
     },
     %{
       host: "b5-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 5,
       port: 9092,
       rack: "use1-az6"
     },
     %{
       host: "b8-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 8,
       port: 9092,
       rack: "use1-az6"
     },
     %{
       host: "b2-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 2,
       port: 9092,
       rack: "use1-az6"
     },
     %{
       host: "b4-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 4,
       port: 9092,
       rack: "use1-az5"
     },
     %{
       host: "b1-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 1,
       port: 9092,
       rack: "use1-az5"
     },
     %{
       host: "b6-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 6,
       port: 9092,
       rack: "use1-az1"
     },
     %{
       host: "b7-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 7,
       port: 9092,
       rack: "use1-az5"
     },
     %{
       host: "b3-<hostname>.us-east-1.aws.confluent.cloud",
       node_id: 3,
       port: 9092,
       rack: "use1-az1"
     }
   ],
   cluster_id: "lkc-8wv98r",
   controller_id: 4,
   topics: [
     %{
       error_code: :no_error,
       is_internal: false,
       name: "test",
       partitions: [
         %{
           error_code: :no_error,
           isr_nodes: [2, 7, 0],
           leader_id: 2,
           partition_index: 0,
           replica_nodes: [2, 7, 0]
         },
         %{
           error_code: :no_error,
           isr_nodes: [7, 3, 8],
           leader_id: 7,
           partition_index: 5,
           replica_nodes: [7, 3, 8]
         },
         %{
           error_code: :no_error,
           isr_nodes: [3, 2, 7],
           leader_id: 3,
           partition_index: 4,
           replica_nodes: [3, 2, 7]
         },
         %{
           error_code: :no_error,
           isr_nodes: [6, 8, 4],
           leader_id: 6,
           partition_index: 1,
           replica_nodes: [6, 8, 4]
         },
         %{
           error_code: :no_error,
           isr_nodes: [1, 6, 5],
           leader_id: 1,
           partition_index: 2,
           replica_nodes: [1, 6, 5]
         },
         %{
           error_code: :no_error,
           isr_nodes: [5, 1, 6],
           leader_id: 5,
           partition_index: 3,
           replica_nodes: [5, 1, 6]
         }
       ]
     }
   ]
 }}
 
 > :brod.produce(:test, "test", 5, "key", "message")
 
17:25:39.750 [info] client :test connected to b7-<hostname>.us-east-1.aws.confluent.cloud:9092

{:ok,
 {:brod_call_ref, #PID<0.893.0>, #PID<0.938.0>,
  #Reference<0.550270668.241696769.28591>}}
```

(to be clear, the call to metadata above is only to display that the correct broker was selected for partition `5`; the client was initialized with `[{"<hostname>.us-east-1.aws.confluent.cloud", 9092}]` as bootstrap endpoints)

**Comment by hairyhum** [comment]
@mszpakowski if we want all connections to brokers to be intercepted by the kafka portal binary we can't allow the application client to connect to something like `b0-<hostname>.us-east-1.aws.confluent.cloud` directly, we want it to connect to the portal binary on `localhost:<port>` so all requests are intercepted.

**Comment by hairyhum** [comment]
In https://github.com/build-trust/ockam/issues/3963#issuecomment-1380725786 we described an ability to use TCP inlet and outlet to connect to kafka brokers and interceptor worker to parse protocol messages.

This means that broker mapping can be done using the protocol messages and dynamic creation of inlets and outlet:

- Each broker returned by the metadata response can be represented by an outlet listener on the outlet side and an inlet listener on the inlet side.
- Inlets can use predefined format to address outlets using node_id from the metadata: `kafka_outlet_<node_id>`
- Outlet interceptor should create an outlet listener for each broker in the metadata response with predefined address format `kafka_outlet_<node_id>`
- Inlet interceptor should create inlets for each broker in the metadata response and point to the right outlet
- Dynamic inlets should route messages through the interceptor worker
- Inlet interceptor should rewrite the metadata response to tell the app client to connect to inlets instead of brokers
- Application should be configured with some port range to start dynamic TCP inlet listeners

 
![Kafka end-to-end - Frame 3](https://user-images.githubusercontent.com/784180/212757133-f32ca5ff-7f41-4f65-96db-1528212be8bd.jpg)

**Comment by hairyhum** [comment]
Addressed in #4155
